arch/xtensa/core/xtensa-win0.S

```diff
@@ -0,0 +1,476 @@
+	wsr \val_reg, LCOUNT
+#endif
+#if XCHAL_HAVE_S32C1I
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_scompare1_OFFSET
+	wsr \val_reg, SCOMPARE1
+#endif
+#if XCHAL_HAVE_THREADPTR
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_threadptr_OFFSET
+	wur \val_reg, THREADPTR
+#endif
+.endm
+
+/* Inc/decrement nested interrupt count on the CPU struct */
+.macro MOD_NEST_COUNT val cpu_reg scratch_reg
+	l32i \scratch_reg, \cpu_reg, ___cpu_t_nested_OFFSET
+	addi \scratch_reg, \scratch_reg, \val
+	s32i \scratch_reg, \cpu_reg, ___cpu_t_nested_OFFSET
+.endm
+
+/* On kernel entry, sets RING to zero (enabling kernel memory access)
+ * and clears EXCM (enabling further nested exceptions and interrupts)
+ */
+.macro PS_ENTRY_UNMASK ps_reg val_reg
+	rsr \ps_reg, PS
+	movi \val_reg, ~(PS_EXCM_MASK | PS_RING_MASK)
+	and \ps_reg, \ps_reg, \val_reg
+	wsr \ps_reg, PS
+.endm
+
+/* Vector handlers: the vector sections themselves are very small (~10
+ * instructions on some devices!), so they do minimal work.  Their job:
+ *
+ * + Place the interrupted a0 register into ZSR_A0SAVE
+ * + Place a C handler address corresponding to the interrupt in a0
+ * + Store the interrupted PC in ZSR_EPC
+ * + Store the interrupted PS register value in ZSR_EPS
+ * + Jump to the shared entry code
+ *
+ * Note that the User and Kernel Level 1 handlers must also check for
+ * two special entry paths (so they do the bulk of their work in a
+ * "l1_handler" section outside the vector region):
+ *
+ * + System calls have their own optimized handler
+ * + MMU TLB exceptions are handled directly, as they must not
+ *   re-enable interrupts at any point before returning to the trapping
+ *   code.  (The handling is very simple: Zephyr does not use "lazy"
+ *   VM regions, and it pins the VM area of the page table itself, so
+ *   the only TLB trap that can happen is a missing entry for the page
+ *   table data, which can be primed into the data TLB by simply
+ *   loading it).
+ */
+.section .UserExceptionVector.text, "ax"
+UserExceptionVector:
+	wsr a0, ZSR_A0SAVE
+	rsr a0, EXCCAUSE
+	beqi a0, EXCCAUSE_SYSCALL, syscall_jump
+	addi a0, a0, -(EXCCAUSE_ITLB_MISS) /* cursed immediate encoding for beqi... */
+	beqi a0, (EXCCAUSE_DTLB_MISS - EXCCAUSE_ITLB_MISS), tlb_jump
+	beqz a0, tlb_jump
+	j l1_handler
+tlb_jump:
+	j tlb_miss
+syscall_jump:
+	j syscall_handler
+
+.section .KernelExceptionVector.text, "ax"
+KernelExceptionVector:
+	wsr a0, ZSR_A0SAVE
+	rsr a0, EXCCAUSE
+	addi a0, a0, -(EXCCAUSE_ITLB_MISS) /* see above */
+	beqi a0, (EXCCAUSE_DTLB_MISS - EXCCAUSE_ITLB_MISS), tlb_miss
+	beqz a0, tlb_miss
+	j l1_handler
+tlb_miss:
+#if XCHAL_HAVE_PTP_MMU
+	rsr a0, PTEVADDR
+	l32i a0, a0, 0
+#endif
+	rsr a0, ZSR_A0SAVE
+	rfe
+
+/* Currently Zephyr uses this only to handle TLB fills in special
+ * circumstances.  Just do that unconditionally.
+ */
+#if XCHAL_HAVE_PTP_MMU
+.section .DoubleExceptionVector.text, "ax"
+_DoubleExceptionVector:
+	wsr a0, ZSR_A0SAVE
+	rsr a0, PTEVADDR
+	l32i a0, a0, 0
```

Comment:
a0 goes nowhere, is anything missing here?

Reply:
This is basically a mis-split.  There was a vestigial handler here in the first version of the PR that just blindly duplicated the existing in-tree handler, which itself was just doing the "load from PTEVADDR to prime the data cache with a page table page" trick (so indeed A0 gets dropped; the side effect on the TLB is what we want).  But that wasn't ever correct to begin with, and a later patch in this PR then replaces it with a jump into the L1 exception handler instead (so later userspace code can handle the fault just like any other memory fault).

There's basically no reason to have this here at all at this point in the PR, I'll remove in the next respin.